samples/solidity/contracts/ERC721WithData.sol

```diff
@@ -76,6 +105,23 @@ contract ERC721WithData is Context, Ownable, ERC721, IERC721WithData {
     }
 
     function _baseURI() internal view virtual override returns (string memory) {
-        return 'firefly://token/';
+        bytes memory tempURITest = bytes(_baseTokenURI);
+        if (tempURITest.length == 0) {
+            return 'firefly://token/';
+        } else {
+            return _baseTokenURI;
+        }
+    }
+
+    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
+        require(_exists(tokenId), "ERC721WithData: Token does not exist");
+
+        string memory uri = _tokenURIs[tokenId];
+        return uri;
+    }
+
+    function _setTokenURI(uint256 tokenId, string memory _tokenURI) internal virtual {
```

Comment:
Was not sure what the permissions around this method should be, if anyone has suggestions.

Reply:
Not sure there's any need to make it virtual

Reply:
a token URI is supposed to be never changed after minting. we can't easily enforce that but requiring `onlyOwner` is close enough